funcbench/main.go

```diff
@@ -0,0 +1,318 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/google/go-github/github"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+type githubClient struct {
+	event                                          *github.IssueCommentEvent
+	client                                         *github.Client
+	owner, repo, branch, latestCommitHash, logLink string
+	prNumber                                       int
+}
+
+type benchmarkHandler struct {
+	bechRegex, home, oldPath, newPath string
+	raceFlagEnabled                   bool
+}
+
+func main() {
+	app := kingpin.New(filepath.Base(os.Args[0]), "benchmark result posting and formating tool.\n-i location of github hook file (even.json)")
+	app.HelpFlag.Short('h')
+	input := app.Flag("input", "path to event.json").Short('i').Default("/github/workflow/event.json").String()
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+
+	data, err := ioutil.ReadFile(*input)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	// Temporary fix for the new Github actions time format, as go-github is not capable of parsing it. This makes the time stamps unusable.
+	txt := string(data)
+	reg := regexp.MustCompile("(.*)\"[0-9]+/[0-9]+/2019 [0-9]+:[0-9]+:[0-9]+ [AP]M(.*)")
+	txt = reg.ReplaceAllString(txt, "$1\"2019-06-11T09:26:28Z$2")
+	data = []byte(txt)
+	// End of the temporary fix.
+
+	event, err := github.ParseWebHook("issue_comment", data)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	switch eventType := event.(type) {
+	case *github.IssueCommentEvent:
+		ghClient := githubClient{event: eventType}
```

Comment:
How about an api that is something like:

```
type gitClient
gitClt,err := newGitclient() // All git setup goes in here - name, email and anyting git related
gitClt.ClonePR
gitClt.CloneBranch

type githubClient
ghClt,err := newGithubclient()
ghClt.PostComment


type benchTest struct {.... 

bench,err := newBencTest(gitClient);
if  err != nil {
	log.Fatalln(err)
}

bench.StartPR(pr)
bench.StartBranch(branch)
bench.Results()
```

Reply:
What are you expecting to be in a function called `gitClt.ClonePR`. If, I assume right, that it is a function, as you have put brackets in only certain parts of your example.

Does your example hint that I should create 3 clients? Also, to what degree is you example a pseudo code? Do you imagine that the main function would contain functions named as in your example and all I would add is error checking?